src/gui/protocolwidget.cpp

```diff
@@ -25,14 +25,115 @@
+    auto f = FolderMan::instance()->folder(folder);
+    if (!f) {
+        return 0;
+    }
+
+    QStringList columns;
+    QDateTime timestamp = QDateTime::currentDateTime();
+    const QString timeStr = timeString(timestamp);
+    const QString longTimeStr = timeString(timestamp, QLocale::LongFormat);
+
+    columns << timeStr;
+    columns << Utility::fileNameForGuiUse(item._originalFile);
+    columns << f->shortGuiLocalPath();
+
+    // If the error string is set, it's prefered because it is a useful user message.
+    QString message = item._errorString;
+    if (message.isEmpty()) {
+        message = Progress::asResultString(item);
+    }
+    columns << message;
+
+    QIcon icon;
+    if (item._status == SyncFileItem::NormalError
+        || item._status == SyncFileItem::FatalError
+        || item._status == SyncFileItem::DetailError
+        || item._status == SyncFileItem::BlacklistedError) {
+        icon = Theme::instance()->syncStateIcon(SyncResult::Error);
+    } else if (Progress::isWarningKind(item._status)) {
+        icon = Theme::instance()->syncStateIcon(SyncResult::Problem);
+    }
+
+    if (ProgressInfo::isSizeDependent(item)) {
+        columns << Utility::octetsToString(item._size);
+    }
+
+    ProtocolItem *twitem = new ProtocolItem(columns);
+    // Warning: The data and tooltips on the columns define an implicit
+    // interface and can only be changed with care.
+    twitem->setData(0, Qt::SizeHintRole, QSize(0, ActivityItemDelegate::rowHeight()));
+    twitem->setData(0, Qt::UserRole, timestamp);
+    twitem->setIcon(0, icon);
+    twitem->setToolTip(0, longTimeStr);
+    twitem->setToolTip(1, item._file);
+    twitem->setData(2, Qt::UserRole, folder);
+    twitem->setToolTip(3, message);
+    twitem->setData(3, Qt::UserRole, item._status);
+    return twitem;
+}
+
+SyncJournalFileRecord ProtocolItem::syncJournalRecord(QTreeWidgetItem *item)
+{
+    SyncJournalFileRecord rec;
+    auto f = folder(item);
+    if (!f)
+        return rec;
+    f->journalDb()->getFileRecord(item->toolTip(1), &rec);
+    return rec;
+}
+
+Folder *ProtocolItem::folder(QTreeWidgetItem *item)
+{
+    return FolderMan::instance()->folder(item->data(2, Qt::UserRole).toString());
+}
+
+void ProtocolItem::openContextMenu(QTreeWidgetItem *item, QWidget *parent)
+{
+    auto f = ProtocolItem::folder(item);
+    if (!f)
+        return;
+    auto rec = ProtocolItem::syncJournalRecord(item);
+    // rec might not be valid
+
+    QObject *noParent = nullptr;
+    QAction openInBrowser(ProtocolWidget::tr("Open in browser"), noParent);
+    QObject::connect(&openInBrowser, &QAction::triggered, parent, [&]() {
+        fetchPrivateLinkUrl(f->accountState()->account(), rec._path,
+            rec.numericFileId(), parent, [parent](const QString &url) {
+                Utility::openBrowser(url, parent);
+            });
+    });
+
+    QMenu menu;
+    if (rec.isValid())
+        menu.addAction(&openInBrowser);
+
+    menu.exec(QCursor::pos());
```

Comment:
Any reason to use QMenu::exec instead of the safer QMenu::popup?
If you capture by value in the lambda, that should work.

Aslo, QCursor::pos() might be wrong, one should use the position given in the slot argument, that's because the menu can be opened with the keyboard shortcut.

Reply:
Using the slot's argument actually placed the menu in a slightly offset location, so I looked at the code we use for popups in other places.

Being asynchronous would be fine (have to take care with `parent` though) and I'll go for that.